db/range_del_aggregator.cc

```diff
@@ -344,22 +344,36 @@ class CollapsedRangeDelMap : public RangeDelMap {
       //     2:    c---   OR   2:    c---   OR   2:    c---   OR   2: c------
       //     1: A--C           1:                1: A------        1: C------
       //                ^                 ^                 ^                  ^
-      // Insert a new transition at the new tombstone's start point, or raise
-      // the existing transition at that point to the new tombstone's seqno.
       end_seq = prev_seq();
-      rep_[t.start_key_] = t.seq_;  // operator[] will overwrite existing entry
+      Rep::iterator pit;
+      if (it != rep_.begin() && (pit = std::prev(it)) != rep_.begin() &&
+          ucmp_->Compare(pit->first, t.start_key_) == 0 && std::prev(pit)->second == t.seq_) {
```

Comment:
This is the grossest part of this fix :/. Taking suggestions on cleaner ways of expressing this.

Reply:
This is exactly the gymnastics I was referring to when I looked at this. Contortions might be a better word.

As a minor cleanup, how about:

```
if (it != rep_.begin() && end_seq == t.seq_ && ucmp_->Compare(std::prev(it)->first, t.start_key_) == 0) {
  ...
}
```

Btw, I'm curious why this is necessary at all. What is blowing up by having two adjacent tombstones with the same sequence number? I saw that we have tests that two adjacent tombstones with different sequence numbers works.

Reply:
> Btw, I'm curious why this is necessary at all. What is blowing up by having two adjacent tombstones with the same sequence number?

The first map in the commit message

a -> 6
b -> 6
c -> 0

is, I think, perfectly valid. It just seemed dumb that we go to all the trouble of building this fancy collapsed map and we can't collapse [a, b) @ 1 and [b, c) @ 1. I think we could make the second map in the commit message

a -> 1
b -> 1
c -> 0
d -> 0

valid as well with a tiny patch like this:

```
diff --git a/db/range_del_aggregator.cc b/db/range_del_aggregator.cc
index cb23a0c6e..e587c59a8 100644
--- a/db/range_del_aggregator.cc
+++ b/db/range_del_aggregator.cc
@@ -150,7 +150,7 @@ class CollapsedRangeDelMap : public RangeDelMap {
 
   class Iterator : public RangeDelIterator {
     void MaybeSeekPastSentinel() {
-      if (Valid() && iter_->second == 0) {
+      while (Valid() && iter_->second == 0) {
         iter_++;
       }
     }
```

But again, seemed like a failure of collapsing.

Re your cleanup suggestion: I don't think that's equivalent? We need to look at the prevprev seq to determine when we're extending an existing tombstone at the same seqno, which I'm not seeing in yours. Unless you're doing something more clever that I didn't consider.

Reply:
Line 347 sets `end_seq = prev_seq();`. So comparing `end_seq == t.seq_` is equivalent to `prev_seq() == t.seq_`.

Reply:
Nope, you're not being dense. I am. I totally missed that subtly. This code is super confusing to read. Hrmm, I don't have a better suggestion right now, though I think you can agree that we should find a clearer structure. Since the tests are passing and this is fixing a real problem, let's merge for now and plan to revisit very soon.